controller/node_controller_test.go

```diff
@@ -1213,6 +1223,101 @@ func (s *NodeControllerSuite) TestCleanupAllInstanceManagers(c *C) {
 	s.checkInstanceManagers(c, expectation)
 }
 
+func (s *NodeControllerSuite) TestNoEventOnUnknownTrueNodeCondition(c *C) {
+	var err error
+
+	node1 := newKubernetesNode(
+		TestNode1,
+		corev1.ConditionTrue,
+		corev1.ConditionFalse,
+		corev1.ConditionFalse,
+		corev1.ConditionFalse,
+		corev1.ConditionFalse,
+		corev1.ConditionFalse,
+		corev1.ConditionTrue,
+	)
+	node1.Status.Conditions = append(node1.Status.Conditions, corev1.NodeCondition{
+		Type:    "EtcdIsVoter",
+		Status:  corev1.ConditionTrue,
+		Reason:  "MemberNotLearner",
+		Message: "Node is a voting member of the etcd cluster",
+	})
+
+	node2 := newKubernetesNode(
+		TestNode2,
+		corev1.ConditionTrue,
+		corev1.ConditionFalse,
+		corev1.ConditionFalse,
+		corev1.ConditionFalse,
+		corev1.ConditionFalse,
+		corev1.ConditionFalse,
+		corev1.ConditionTrue,
+	)
```

Comment:
Suggest covering all checked conditions.

[/controller/node_controller.go#L437-L464](https://github.com/longhorn/longhorn-manager/blob/c46a5456aacf3441f40eab70d95fc7d58ab230de/controller/node_controller.go#L437-L464)
```go
			case corev1.NodeReady:
				if con.Status != corev1.ConditionTrue {
					if con.Status == corev1.ConditionFalse &&
						time.Since(con.LastTransitionTime.Time) < ignoreKubeletNotReadyTime {
						// When kubelet restarts, it briefly reports Ready == False. Responding too quickly can cause
						// undesirable churn. See https://github.com/longhorn/longhorn/issues/7302 for an example.
						nc.logger.Warnf("Ignoring %v == %v condition due to %v until %v", corev1.NodeReady, con.Status,
							con.Reason, con.LastTransitionTime.Add(ignoreKubeletNotReadyTime))
					} else {
						node.Status.Conditions = types.SetConditionAndRecord(node.Status.Conditions,
							longhorn.NodeConditionTypeReady, longhorn.ConditionStatusFalse,
							string(longhorn.NodeConditionReasonKubernetesNodeNotReady),
							fmt.Sprintf("Kubernetes node %v not ready: %v", node.Name, con.Reason),
							nc.eventRecorder, node, corev1.EventTypeWarning)
					}
				}
			case corev1.NodeDiskPressure,
				corev1.NodePIDPressure,
				corev1.NodeMemoryPressure,
				corev1.NodeNetworkUnavailable:
				if con.Status == corev1.ConditionTrue {
					node.Status.Conditions = types.SetConditionAndRecord(node.Status.Conditions,
						longhorn.NodeConditionTypeReady, longhorn.ConditionStatusFalse,
						string(longhorn.NodeConditionReasonKubernetesNodePressure),
						fmt.Sprintf("Kubernetes node %v has pressure: %v, %v", node.Name, con.Reason, con.Message),
						nc.eventRecorder, node, corev1.EventTypeWarning)
				}
			}
```

Reply:
Looks like they are all already covered? https://github.com/m-ildefons/longhorn-manager/blob/b810121b33789d145f220bfd0e41102a7801a354/controller/volume_controller_test.go#L1250-L1280

Reply:
I think what David is referring to here is that we should have tests for each known node condition to make sure if events are emitted, they are exactly as expected. I.e. if there is a `MemoryPressure`, `DiskPressure`, etc. condition that is true, that we do see the corresponding "Kubernetest node X has pressure" event emitted by the Longhorn Manager. Likewise check that the manager emits events for the `NodeReady` status condition as expected as well.
This is not yet covered by the tests and there are other controllers that also emit events.